modules/stochastic_tools/src/samplers/AIS.C

```diff
@@ -0,0 +1,177 @@
+  _prev_value.resize(_distributions.size());
+
+  // `check_step` is a member variable for ensuring that the MCMC algorithm proceeds in a sequential
+  // fashion.
+  _check_step = 0;
+
+  // Storage for means of input values for proposing the next sample
+  _mean_sto.resize(_distributions.size());
+
+  // Storage for standard deviations of input values for proposing the next sample
+  _std_sto.resize(_distributions.size());
+
+  setNumberOfRandomSeeds(_num_random_seeds);
+}
+
+Real
+AIS::computeSample(dof_id_type /*row_index*/, dof_id_type col_index)
+{
+  TIME_SECTION(_perf_compute_sample);
+
+  if (_step <= _num_samples_train)
+  {
+    /* This is the importance distribution training step. Markov Chains are set up
+       to sample from the importance region or the failure region using the Metropolis
+       algorithm. Given that the previous sample resulted in a model failure, the next
+       sample is proposed such that it is very likely to result in a model failure as well.
+       The `initial_values` and `proposal_std` parameters provided by the user affects the
+       formation of the importance distribution. */
+    if (_step > 1 && col_index == 0 && _check_step != _step)
+    {
+      for (dof_id_type j = 0; j < _distributions.size(); ++j)
+        _prev_value[j] = Normal::quantile(
+            _distributions[j]->cdf(getReporterValueByName<Real>(_inputs_names[j])), 0, 1);
+      _acceptance_ratio = 0.0;
+      for (dof_id_type i = 0; i < _distributions.size(); ++i)
+        _acceptance_ratio += std::log(Normal::pdf(_prev_value[i], 0, 1)) -
+                             std::log(Normal::pdf(_inputs_sto[i][_inputs_sto[i].size() - 1], 0, 1));
+      if (_acceptance_ratio > std::log(getRand(_step)))
+      {
+        for (dof_id_type i = 0; i < _distributions.size(); ++i)
+          _inputs_sto[i].push_back(_prev_value[i]);
+      }
+      else
+      {
+        for (dof_id_type i = 0; i < _distributions.size(); ++i)
+          _inputs_sto[i].push_back(_inputs_sto[i][_inputs_sto[i].size() - 1]);
+      }
+    }
+    _check_step = _step;
+    _prev_value[col_index] =
+        Normal::quantile(getRand(_step),
+                         _inputs_sto[col_index][_inputs_sto[col_index].size() - 1],
+                         _proposal_std[col_index]);
+    return _distributions[col_index]->quantile(Normal::cdf(_prev_value[col_index], 0, 1));
+  }
+  else if (_step == _num_samples_train + 1)
+  {
+    /* This is the importance sampling step using the importance distribution created
+       in the previous step. Once the importance distribution is known, sampling from
+       it is similar to a regular Monte Carlo sampling. */
+    if (col_index == 0 && _check_step != _step)
+    {
+      for (dof_id_type i = 0; i < _distributions.size(); ++i)
+      {
+        _mean_sto[i] = StochasticTools::AdaptiveMonteCarloUtils::computeMEAN(_inputs_sto[i], 2);
+        _std_sto[i] = StochasticTools::AdaptiveMonteCarloUtils::computeSTD(_inputs_sto[i], 2);
```

Comment:
These are pretty simple functions that I don't think you need utilities for. For instance, these can be changed to:
```cpp
const Real count = _inputs_sto.size() - 2;
const Real mean = std::accumulate(_inputs_sto[i].begin() + 2, _inputs_sto[i].end(), 0.0) / count;
const Real var = std::accumulate(_inputs_sto[i].begin() + 2, _inputs_sto[i].end(), 0.0,
                          [&mean](Real x, Real y) { return x + (y - mean) * (y - mean); }) / count;
_mean_sto[i] = mean;
_std_sto[i] = std::sqrt(var);
```

Reply:
I have another algorithm called Subset Simulation which requires the above two functions as wells as new functions such as returning the largest po fraction of a vector (0<=po<=1). In addition, other adaptive Monte Carlo algs would require such functions as well. This is why I created the utils class to store all such functions for adaptive Monte Carlo at one place and avoid code duplication. Let me know what you think.